File: quicklib/src/main/java/com/sdwfqin/quicklib/mvvm/BaseViewModel.java
package com.sdwfqin.quicklib.mvvm;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

/**
 * BaseViewModel
 * <p>
 *
 * @author 张钦
 * @date 2020/4/15
 */
public class BaseViewModel extends ViewModel {

    /**
     * 加载窗状态
     */
    public final MutableLiveData<Boolean> isLoading = new MutableLiveData<>();

    /**
     * 通用网络请求异常
     */
    public final MutableLiveData<Throwable> networkError = new MutableLiveData<>();
}

File: quicklib/src/main/java/io/github/sdwfqin/quicklib/mvvm/BaseViewModel.kt
package io.github.sdwfqin.quicklib.mvvm

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch

/**
 * BaseViewModel
 *
 * @author 张钦
 * @date 2020/4/15
 */
open class BaseViewModel : ViewModel() {
    /**
     * 加载窗状态
     */
    val eventLoading = MutableSharedFlow<Boolean>()

    /**
     * 通用网络请求异常
     */
    val eventError = MutableSharedFlow<Throwable>()

    fun launch(
        block: suspend () -> Unit,
        error: suspend (Throwable) -> Boolean = { true },
        complete: suspend () -> Boolean = { true }
    ) = viewModelScope.launch {
        try {
            block()
        } catch (e: Throwable) {
            if (error(e)) {
                eventError.emit(e)
            }
        } finally {
            if (complete()) {
                eventLoading.emit(false)
            }
        }
    }
}